Dual write to OSS and Cloud (#3383)
influxdata/docs-v2: content/influxdb/v2.1/write-data/no-code/use-telegraf/dual-write.md

---
title: Dual write to InfluxDB OSS and InfluxDB Cloud
description: Write data to both OSS and Cloud simultaneously.
menu:
  influxdb_2_1:
    parent: Telegraf (agent)
weight: 201
---

Write to both InfluxDB OSS and InfluxDB Cloud with Telegraf.

You might want to set up dual write if you want a backup of your data in two places, or if you're migrating from OSS to Cloud.

The sample configuration below uses the [InfluxDB v2 output plugin](https://github.com/influxdata/telegraf/tree/master/plugins/outputs/influxdb_v2) twice: first pointing to the OSS instance and then to the cloud instance.

## Sample configuration

```
[[inputs.cpu]]
[[inputs.mem]]

## Any other inputs, processors, aggregators that you want to include into your config.

> ## Without any namepass, tagpass filters, or settings changes, your data should get identically written to the two outputs below.

Reviewer: Can probably move this line out of the sample config actually and into the description. Can even change it to "Without any [metric filtering](https://docs.influxdata.com/telegraf/latest/administration/configuration/#metric-filtering) or [setting changes](https://docs.influxdata.com/telegraf/v1.20/administration/configuration/#output-configuration) to your output plugins, your data should get identically written to your OSS and Cloud instances"

# Send data to InfluxDB OSS v2
[[outputs.influxdb_v2]]
  ## The URLs of the InfluxDB cluster nodes.
  ##
  ## Multiple URLs can be specified for a single cluster, only ONE of the
  ## urls will be written to each interval.
  ## urls exp: http://127.0.0.1:9999
  urls = ["http://localhost:8086"]

  ## Token for authentication.
  token = "$INFLUX_TOKEN"

  ## Organization is the name of the organization you want to write to. It must already exist.
  organization = "influxdata"

  ## Destination bucket to write into.
  bucket = "telegraf"

# Send data to InfluxDB Cloud - AWS West cloud instance
 [[outputs.influxdb_v2]]
  ## The URLs of the InfluxDB Cloud cluster nodes.

  urls = ["https://us-west-2-1.aws.cloud2.influxdata.com"]

  ## Token for authentication.

>   token = "$INFLUX_TOKEN"

Reviewer 2: two different Tokens are required....which means you need to have two different environment variables to leverage a specific token for each target.

  ## Organization is the name of the organization you want to write to. It must already exist.

>   organization = "swang@influxdata.com"

Reviewer 2: let's not use this specific email address in the PR.... I would prefer we just use an example email address.  like.... `example@my_email.com`

Reviewer: haha yes please. don't need any more e-mail spam 😝

---
title: Dual write to InfluxDB OSS and InfluxDB Cloud
description: Write data to both OSS and Cloud simultaneously.
menu:
  influxdb_2_1:
    parent: Telegraf (agent)
weight: 201
---

> Write to both InfluxDB OSS and InfluxDB Cloud with Telegraf.
> 
> You might want to set up dual write if you want a backup of your data in two places, or if you're migrating from OSS to Cloud.

Reviewer 3: ```suggestion
If you want to back up your data in two places, or if you're migrating from OSS to Cloud, you may want to set up dual write. 

Use Telegraf to write to both InfluxDB OSS and InfluxDB Cloud simuiltaneously.
```

The sample configuration below uses:
  - The [InfluxDB v2 output plugin](https://github.com/influxdata/telegraf/tree/master/plugins/outputs/influxdb_v2) twice: first pointing to the OSS instance and then to the cloud instance.

>   - Two different tokens, one for OSS and one for Cloud. You'll need to configure both tokens as environment variables (see [Configure your token as an environment variable](/influxdb/{{< latest "influxdb" "v2" >}}/write-data/no-code/use-telegraf/auto-config/#configure-your-token-as-an-environment-variable)).

Reviewer 3: ```suggestion
  - Two different tokens, one for OSS and one for Cloud. You'll need to configure both tokens as environment variables (see [Configure your token as an environment variable](/{{< latest "influxdb" "v2" >}}/write-data/no-code/use-telegraf/auto-config/#configure-your-token-as-an-environment-variable)).
```

Reviewer 3: please double-check to be sure this link works

> Without any metric filtering or setting changes to your output plugins, the configuration below writes your data identically to your OSS and Cloud instances.

Reviewer 3: ```suggestion
Use the configuration below to write your data to both OSS and Cloud instances simultaneously. 
```

Reviewer 3: if we feel it's important to highlight this part: Without any metric filtering or setting changes to your output plugins
...would this be accurate?: Without filtering metrics or changing your output plugins

or something more parallel

Reviewer: i think it would be specifically adjusting output plugin behavior. "changing output plugins" is kinda vague

## Sample configuration

```
[[inputs.cpu]]
[[inputs.mem]]

> ## Any other inputs, processors, aggregators that you want to include into your config.

Reviewer 3: ```suggestion
## Any other inputs, processors, aggregators that you want to include in your configuration.
```

# Send data to InfluxDB OSS v2
[[outputs.influxdb_v2]]

>   ## The URLs of the InfluxDB cluster nodes.

Reviewer 3: are cluster nodes accurate for OSS?

Reviewer: mm not really. I would just say "The URLs of the InfluxDB instance"

  ##
  ## Multiple URLs can be specified for a single cluster, only ONE of the
  ## urls will be written to each interval.
  ## urls exp: http://127.0.0.1:9999
  urls = ["http://localhost:8086"]

  ## OSS token for authentication.
  token = "$INFLUX_TOKEN_OSS"

  ## Organization is the name of the organization you want to write to. It must already exist.
  organization = "influxdata"

>   ## Destination bucket to write into.

Reviewer 3: ```suggestion
  ## Destination bucket to write to.
```
